### PR TITLE
fix(metrics): score a missed prediction in Hausdorff percentile, don't drop it

### LINKED ISSUE #9095: HausdorffDistanceMetric percentile returns NaN for a missed prediction, dropping failures from the dataset average
**Describe the bug**

`HausdorffDistanceMetric(percentile=...)` returns `nan` when one of the two masks is empty, where `percentile=None` returns `inf` for the same input.

`nan` is not a quieter way of saying `inf` here. It is this metric's "not applicable" sentinel — it is what both-masks-empty returns — and `do_metric_reduction` excludes it from the average. A prediction that missed the structure entirely is therefore removed from a dataset score rather than counted as the worst case, and the reported HD95 improves as the model finds fewer structures.

**To Reproduce**

```python
import torch
from monai.metrics import HausdorffDistanceMetric

gt = torch.zeros(1, 1, 64, 64)
gt[0, 0, 20:28, 20:28] = 1.0
pred = torch.zeros(1, 1, 64, 64)          # the model found nothing

for percentile in (None, 50, 95, 99, 100):
    m = HausdorffDistanceMetric(include_background=True, percentile=percentile)
    m(y_pred=pred, y=gt)
    print(percentile, m.get_buffer().flatten().tolist())

# None -> [inf]
# 50   -> [nan]
# 95   -> [nan]
# 99   -> [nan]
# 100  -> [nan]
```

The effect over a dataset. 100 images, every non-empty prediction identically 4px off, so the only variable is how many images the model misses:

| images missed | reported HD95 | cases averaged (`get_not_nans`) |
| ---: | ---: | ---: |
| 0 | 4.000 | 100 |
| 50 | 4.000 | 50 |
| 90 | 4.000 | 10 |
| 99 | 4.000 | 1 |

A model that finds nothing in 99 images out of 100 reports the same HD95 as one that finds the structure every time. A runnable script that prints the whole table is at https://github.com/asifuddin01/A-PR.

**Expected behavior**

`inf`, matching `percentile=None`, so the case is scored rather than discarded. `nan` stays reserved for both masks empty, where there is genuinely nothing to measure.

**Cause**

`get_surface_distance` reports an infinite distance for every boundary voxel when a mask is empty, so an all-infinite tensor reaches `_compute_percentile_hausdorff_distance`. `torch.quantile` interpolates linearly between the two order statistics straddling the requested rank:

```
inf + (inf - inf) * 0.95  ->  nan
```

The quantile of a constant sequence is that constant, so the `nan` is an artefact of the interpolation rather than a property of the distances. `percentile=None` escapes it because `.max()` does not interpolate.

**Environment**

```
MONAI version: 1.6.0
Numpy version: 2.5.2
Pytorch version: 2.14.0
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: eccefc57550b111ed781d82249dfe77872a0e918
```

Also reproduced on current `dev` at c0d1ec1.

**Additional context**

Only the percentile path is affected. `SurfaceDistanceMetric` returns `inf` for the same input and `SurfaceDiceMetric` returns `0.0`; both are already honest about a miss.

Separately, #9033 fixes `percentile=0` being treated as unset by the `if not percentile:` guard. It does not cover this: it adds a `.min()` branch above the `torch.quantile` call and leaves the interpolation as it is. The two are independent.

I have a fix and regression tests ready and will open a PR against this issue.

### PR DESCRIPTION
**Pull request by asifuddin01**
Fixes #9095.

### Description

`HausdorffDistanceMetric(percentile=...)` returns `nan` when one of the two masks is empty, where `percentile=None` returns `inf` for the same input.

`get_surface_distance` reports an infinite distance for every boundary voxel when a mask is empty, so an all-infinite tensor reaches `_compute_percentile_hausdorff_distance`. `torch.quantile` interpolates linearly between the two order statistics straddling the requested rank, and that interpolation is `inf + (inf - inf) * frac`, which is `nan`. The maximum and minimum paths escape it because they do not interpolate.

The quantile of a constant sequence is that constant, so the `nan` is an artefact rather than a property of the distances. It also collides with the meaning this metric already gives `nan`: both-masks-empty returns it to say "not applicable", and `do_metric_reduction` excludes it from the average. A prediction that missed the structure entirely is therefore removed from a dataset score rather than counted as the worst case, and the reported HD95 improves as the model finds fewer structures. `get_not_nans=True` exposes the shrinking denominator but is off by default. The issue has the table.

This returns the infinity directly when every distance is infinite, so the percentile path agrees with the maximum path:

```python
if torch.isinf(surface_distance).all():
    return torch.tensor(np.inf, dtype=torch.float, device=surface_distance.device)
return torch.quantile(surface_distance, percentile / 100)
```

The guard is exact rather than defensive. `get_surface_distance` returns either all-finite or all-infinite distances and never a mixture, because the infinite branch triggers on an empty mask, which makes every distance infinite at once; each direction of the symmetric distance is computed and reduced separately, so a partly-infinite tensor does not reach this function.

Tests cover both entry points, the metric class and the helper, at `percentile` None, 0, 50, 95, 99 and 100. With the source change reverted and the tests kept, eight fail and four pass, the four being None and 0, which take the non-interpolating paths.

On testing: I ran `tests/metrics/` in full (393 passed, 41 skipped) and `tests/metrics/test_hausdorff_distance.py` (68 passed), plus black, isort and ruff on both changed files. One pre-existing failure in `test_compute_fid_metric.py` reproduces on unmodified `dev` with torch 2.14 and is unrelated to this change. I have not completed a full `./runtests.sh --quick --unittests` run locally, so I have left those boxes unticked rather than tick them untested.

Related but independent: #9033 fixes `percentile=0` being treated as unset by the `if not percentile:` guard. It adds a `.min()` branch above the `torch.quantile` call and leaves the interpolation as it is, so the two changes do not overlap in behaviour. They touch nearby lines and whichever lands second will want a trivial rebase; happy to be the one that rebases.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by asifuddin01** [comment]
Addressed the docstring note in 9b168f4: Google-style `Returns` and `Raises` on `compute_hausdorff_distance`, and `Args`/`Returns`/`Raises` on `_compute_percentile_hausdorff_distance`, which had a one-line docstring before. They now state the distinction this PR turns on, which was previously only inferable from the code: `nan` when neither mask has a foreground, so there is nothing to measure and the reduction excludes it, and `inf` when exactly one mask is empty, at the maximum and at every percentile alike.

For whoever picks this up: `tests/metrics/` passes locally, 434 passed. The one failure there, `test_compute_fid_metric.py::TestFIDMetric::test_results`, reproduces on unmodified `dev` with torch 2.14 and is unrelated to this change.

The workflows are sitting at `action_required` as a first-time contributor, so they need an approval to run.
